src/validation/validation.cpp

```diff
@@ -2161,33 +2154,248 @@ bool ConnectBlock(const CBlock &block,
+    std::vector<uint256> &vHashesToDelete)
+{
+    nFees = 0;
+    int64_t nTime2 = GetTimeMicros();
+    LOG(BLK, "Canonical ordering for %s MTP: %d\n", block.GetHash().ToString(), pindex->GetMedianTimePast());
+
+    // Start enforcing BIP68 (sequence locks) and BIP112 (CHECKSEQUENCEVERIFY) using versionbits logic.
+    int nLockTimeFlags = 0;
+    if (VersionBitsState(pindex->pprev, chainparams.GetConsensus(), Consensus::DEPLOYMENT_CSV, versionbitscache) ==
+        THRESHOLD_ACTIVE)
     {
-        return false; // no need to lock cs_main before returning as it should already be locked.
+        nLockTimeFlags |= LOCKTIME_VERIFY_SEQUENCE;
     }
 
-    // Quit any competing threads may be validating which have the same previous block before updating the UTXO.
-    PV->QuitCompetingThreads(block.GetBlockHeader().hashPrevBlock);
+    // Get the script flags for this block
+    uint32_t flags = GetBlockScriptFlags(pindex, chainparams.GetConsensus());
+    bool fStrictPayToScriptHash = flags & SCRIPT_VERIFY_P2SH;
+
+    ValidationResourceTracker resourceTracker;
+    std::vector<int> prevheights;
+    int nInputs = 0;
+    unsigned int nSigOps = 0;
+    CDiskTxPos pos(pindex->GetBlockPos(), GetSizeOfCompactSize(block.vtx.size()));
+    blockundo.vtxundo.reserve(block.vtx.size() - 1);
+    int nChecked = 0;
+    int nOrphansChecked = 0;
+    const arith_uint256 nStartingChainWork = chainActive.Tip()->nChainWork;
+
+    // Section for boost scoped lock on the scriptcheck_mutex
+    boost::thread::id this_id(boost::this_thread::get_id());
+
+    // Get the next available mutex and the associated scriptcheckqueue. Then lock this thread
+    // with the mutex so that the checking of inputs can be done with the chosen scriptcheckqueue.
+    CCheckQueue<CScriptCheck> *pScriptQueue(PV->GetScriptCheckQueue());
+
+    // Aquire the control that is used to wait for the script threads to finish. Do this after aquiring the
+    // scoped lock to ensure the scriptqueue is free and available.
+    CCheckQueueControl<CScriptCheck> control(fScriptChecks && PV->ThreadCount() ? pScriptQueue : NULL);
+
+    // Initialize a PV session.
+    if (!PV->Initialize(this_id, pindex, fParallel))
+        return false;
+
+    /*********************************************************************************************
+     If in PV, unlock cs_main here so we have no contention when we're checking inputs and scripts
+     *********************************************************************************************/
+    if (fParallel)
+        LEAVE_CRITICAL_SECTION(cs_main);
+
+    // Begin Section for Boost Scope Guard
+    {
+        // Scope guard to make sure cs_main is set and resources released if we encounter an exception.
+        BOOST_SCOPE_EXIT(&fParallel) { PV->SetLocks(fParallel); }
+        BOOST_SCOPE_EXIT_END
+
+        // Outputs then Inputs algorithm: add outputs to the coin cache
+        // and validate lexical ordering
+        uint256 prevTxHash;
```

Comment:
Don't we just need this "add outputs first" loop in ConnectBlock() when the fork is activated and not this entire code block, seems like a lot of duplicate code. 

Other than this last comment I'm done my review and the tests seem reliable now.